### ISSUE TITLE
Import Pothos module fails

### ISSUE THREAD

**Issue by TheJimD** [pothosware/pothos-python#9]
Hi,

I installed Pothosware using the Windows installer several weeks ago and have been using the blocks with my RTL-SDR.  That's been going well so I wanted to try creating some of my own Python processing blocks.  I'm trying to start with the Adder example to get the basics down.  

When I try to import Pothos, I get an ImportError - "no module named Pothos". I checked my sys.path and it includes C:\\Program Files\\PothosSDR\\lib\\site-packages\\Pothos (which I think is the correct folder).  I am using the Anaconda Python 2.7 distribution and the Spyder IDE create & run my Python code.  Other Python imports (like numpy) work just fine.

I had a similar issue when I tried to import Soapy-SDR. So I think I am missing something very basic here.  Do I need to download and build the Pothos Python from source?  I have  CMake and Visual Studio 15.  I am pretty new at using CMake so instructions would be helpful. :)

Thanks,
   Jim Davis

**Comment by guruofquality** [comment]
The python bindings are built for python 3.4, so you will need the equivalent python 3.4 environment and modules. It would be possible in the future to have support for both 3.4 and 2.7, although currently the build for SoapySDR and pothos-python would need some tweaks to avoid using the same install directory (making a issue on that now).

* Some more information here: https://github.com/pothosware/PothosSDR/wiki/Development#python-processing-blocks
* There is a basic example here: https://github.com/pothosware/pothos-python/wiki#a-simple-example
* And the cmake + build instructions are the same as a C++ block (I do recommend using the cmake-gui that comes with the cmake) https://github.com/pothosware/pothos/wiki/BlocksCodingGuide#building-and-installing

**Comment by TheJimD** [comment]
Ah - I missed the part about needing python 3.4.  I had installed 2.7 a couple of years ago to fool around with as we were migrating from Matlab to Python at work.  I'm updating to Anaconda3 version 2.5.0 which says it supports python 3.4 & 3.5.  I probably don't need everything that Anaconda installs but it's convenient to just install one thing.

I will try that and see how it goes.  

Thanks for the help,
  Jim

**Comment by TheJimD** [comment]
So I have made some good progress today.  After installing Anaconda3, I set up a working area and downloaded the pothos-python repository. I used CMakeGUI to configure & generate for Visual Studio 15 and then built the RelWithDebInfo 64-bit version.

I copied the __init__.py, PothosModule.pyd, and 6 more .py files (Block, Buffer, InputPort, Label, & OutputPort) into a temporary lib\site_packages folder in my working area.  When I include that folder in my PythonPath, I was able to import PothosModule into Spyder.  Since that was failing before, I think this means I was successful!!

I subsequently found a cmake_install.cmake file and was able to use that (by running cmake -P cmake_install.cmake as admin in a command window) to copy the PothosModule.pyd and other files to the C:\ProgramFiles\PothosSDR\Lib\site-packages\Pothos folder.  I removed my temporary working folder from the PythonPath, restarted Spyder, and was able to import PothosModule.

Next I will try to create the Adder block from the basic example and see where that leads me.

I suspect I can follow a similar process for building the SoapySDR repository so that I can import it into Python also.

CMake is still sort of a mystery to me but I think I am starting to understand it.  

Jim D

**Comment by guruofquality** [comment]
> I used CMakeGUI to configure & generate for Visual Studio 15 and then built the RelWithDebInfo 64-bit version.

Just for a data point. What version of python were you building against?

> I subsequently found a cmake_install.cmake file and was able to use that (by running cmake -P cmake_install.cmake as admin in a command window) to copy the PothosModule.pyd and other files to the C:\ProgramFiles\PothosSDR\Lib\site-packages\Pothos folder. I removed my temporary working folder from the PythonPath, restarted Spyder, and was able to import PothosModule.

Cool. Something that comes up is installing modules to directories outside of the main installer root (for permissions or other reasons). So if it helps, the POTHOS_PLUGIN_PATH and PYTHONPATH environment variables can be set to include external projects.

```
set POTHOS_PLUGIN_PATH=c:\myapp\lib\Pothos\modules
set PYTHONPATH=c:\myapp\lib\site_packages
```

FWIW, I gave myself some issues to better support multiple versions of python in the installer, building for 3.5 and installer setting the path automatically:

* https://github.com/pothosware/pothos-python/issues/10
* https://github.com/pothosware/SoapySDR/issues/62
* https://github.com/pothosware/PothosSDR/issues/18
* https://github.com/pothosware/PothosSDR/issues/19

There should be no problem building SoapySDR simultaneously for every version of python, but the pothos-python bindings are a bit interesting because of how it calls into python from the library side. It looks like simultaneous versions of python runtime dont want to be loaded into memory at the same time (they step on each other). Otherwise, you could write blocks, each under a different python version, and it would interoperate. As it stands, I will need a way to let user's pick the python version somehow which will select the matching bindings library.

**Comment by TheJimD** [comment]
The Python version I'm using is 3.5.1 from the Anaconda3 version 2.5.0 64-bit Windows distribution.  I also discovered that I could enable the Python Tools for Visual Studio so I have done that and may move away from the Spyder IDE to just using Visual Studio for C & Python.

I got busy over the rest of the weekend and haven't gotten back to trying the Adder example.  But now that I can import the PythosModule successfully I should be able to work through that when I get some time.

All of these mixed languages and third party tools are a little overwhelming for a guy who learned to program on punch cards in the late '70s.  :)  I don't get to do much programming at work any more so this is a fun exercise for me.

Thanks again for the assistance,
   Jim
